{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "55ad06d4-b76b-4f6f-88b5-6a4db236739c",
   "metadata": {},
   "source": [
    "# Day 4 Part 1\n",
    "---\n",
    "\n",
    "## Multiple sequence aligners + Phylogenetics\n",
    "\n",
    "For more information about the theory behind of multiple sequence alignment, please see these resources:\n",
    "\n",
    "- https://academic.oup.com/bib/article/17/6/1009/2606431\n",
    "- https://www.hindawi.com/journals/isrn/2013/615630/\n",
    "\n",
    "Note: There is a user-friend tool called MEGA that is likely sufficient for analyzing closely related species (within a phylum). However, it lacks some of the more sophisticated phylogenetic methods https://www.megasoftware.net/\n",
    "\n",
    "### Aligners\n",
    "\n",
    "Here is a list of recommended aligners (non-exhaustive). Unfortunately, most do not have Windows packages, however some of them do have webservers.  \n",
    "- [MAFFT](https://mafft.cbrc.jp/alignment/software/) \n",
    "- [T-COFFEE](http://www.tcoffee.org/Projects/tcoffee/workshops/tcoffeetutorials/installation.html)\n",
    "- [MUSCLE](https://www.drive5.com/muscle/) -- Has a windows distribution!\n",
    "- [HMMER](http://hmmer.org/) *great for aligning distantly related proteins*\n",
    "  - Profile alignment, guides the alignment of your sequences using a well curated template file. \n",
    "\n",
    "### Alignment viewers\n",
    "\n",
    "The most popular viewer on the market is [Jalview](https://www.jalview.org/), but there are older alternatives such as [SeaView](http://doua.prabi.fr/software/seaview).\n",
    "\n",
    "### Masking alignments\n",
    "\n",
    "If you want to proceed to phylogenetic analysis you can treat your alignment with a 'masking' method. This is a way to 'mask' or 'hide' regions of the alignment that might be misaligned (and therefore non-homologous). Historically, this was done by eye (yep..) but now we have tools to help us out! My favourites are: \n",
    "\n",
    "- [BMGE](https://bmcecolevol.biomedcentral.com/articles/10.1186/1471-2148-10-210)\n",
    "- [trimal](http://trimal.cgenomics.org/) -- has a windows distribution\n",
    "- [divvier](https://github.com/simonwhelan/Divvier) # advanced user\n",
    "\n",
    "### Phylogenetic Analyses\n",
    "\n",
    "Finally, for phylogenetic analyses, I recommend Maximum-likelihood or Bayesian methods:\n",
    "\n",
    "- [IQTREE](http://www.iqtree.org/)\n",
    "- [RAxML](https://cme.h-its.org/exelixis/web/software/raxml/)\n",
    "- [PhyloBayes](http://www.atgc-montpellier.fr/phylobayes/)\n",
    "\n",
    "### TreeViewers\n",
    "\n",
    "- [FIGTREE](http://tree.bio.ed.ac.uk/software/figtree/)\n",
    "- [iTOL](https://itol.embl.de/)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7bb5a53-7f33-4e09-a7a2-cdfbbcf2a226",
   "metadata": {},
   "source": [
    "Since muscle has a windows distribution let's try and get that working.  Mac/unix/colab users head to the next section\n",
    "\n",
    "### Windows\n",
    "\n",
    "Window's folks head to this link and download the executable (tar.gz again): \n",
    "\n",
    "https://www.drive5.com/muscle/downloads.htm\n",
    "\n",
    "http://trimal.cgenomics.org/downloads\n",
    "\n",
    "http://www.iqtree.org/#download\n",
    "\n",
    "\n",
    "Note the path where you have downlodaded these including the \\bin\\\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06f2b420-be4f-4554-afdb-c452095e9bd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the path variable for each package\n",
    "!set PATH=%PATH%;C:\\your\\path\\here\\\n",
    "!muscle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9726eb2b-25fc-43b6-ab12-0e0bdc3ab09d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/julieb/anaconda3\n",
      "\n",
      "  added / updated specs:\n",
      "    - muscle\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    conda-4.14.0               |   py39hf3d152e_0        1011 KB  conda-forge\n",
      "    muscle-3.8.1551            |       h7d875b9_6         262 KB  bioconda\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:         1.2 MB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  muscle             bioconda/linux-64::muscle-3.8.1551-h7d875b9_6\n",
      "\n",
      "The following packages will be UPDATED:\n",
      "\n",
      "  conda                               4.13.0-py39hf3d152e_1 --> 4.14.0-py39hf3d152e_0\n",
      "\n",
      "\n",
      "\n",
      "Downloading and Extracting Packages\n",
      "conda-4.14.0         | 1011 KB   | ##################################### | 100% \n",
      "muscle-3.8.1551      | 262 KB    | ##################################### | 100% \n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n",
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/julieb/anaconda3\n",
      "\n",
      "  added / updated specs:\n",
      "    - trimal\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    trimal-1.4.1               |       h7d875b9_5         189 KB  bioconda\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:         189 KB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  trimal             bioconda/linux-64::trimal-1.4.1-h7d875b9_5\n",
      "\n",
      "\n",
      "\n",
      "Downloading and Extracting Packages\n",
      "trimal-1.4.1         | 189 KB    | ##################################### | 100% \n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n",
      "Retrieving notices: ...working... done\n",
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/julieb/anaconda3\n",
      "\n",
      "  added / updated specs:\n",
      "    - iqtree\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    iqtree-2.0.3               |       h176a8bc_1         3.0 MB  bioconda\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:         3.0 MB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  iqtree             bioconda/linux-64::iqtree-2.0.3-h176a8bc_1\n",
      "\n",
      "\n",
      "\n",
      "Downloading and Extracting Packages\n",
      "iqtree-2.0.3         | 3.0 MB    | ##################################### | 100% \n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n",
      "Retrieving notices: ...working... done\n"
     ]
    }
   ],
   "source": [
    "### MAC/UNIX/COLLAB\n",
    "\n",
    "!conda install --yes -c bioconda muscle\n",
    "!conda install -c bioconda --yes trimal\n",
    "!conda install --yes -c bioconda iqtree\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "32f2625e-0d6f-4369-ba39-68bd9488a253",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "\n",
      "Basic usage\n",
      "\n",
      "    muscle -in <inputfile> -out <outputfile>\n",
      "\n",
      "Common options (for a complete list please see the User Guide):\n",
      "\n",
      "    -in <inputfile>    Input file in FASTA format (default stdin)\n",
      "    -out <outputfile>  Output alignment in FASTA format (default stdout)\n",
      "    -diags             Find diagonals (faster for similar sequences)\n",
      "    -maxiters <n>      Maximum number of iterations (integer, default 16)\n",
      "    -maxhours <h>      Maximum time to iterate in hours (default no limit)\n",
      "    -html              Write output in HTML format (default FASTA)\n",
      "    -msf               Write output in GCG MSF format (default FASTA)\n",
      "    -clw               Write output in CLUSTALW format (default FASTA)\n",
      "    -clwstrict         As -clw, with 'CLUSTAL W (1.81)' header\n",
      "    -log[a] <logfile>  Log to file (append if -loga, overwrite if -log)\n",
      "    -quiet             Do not write progress messages to stderr\n",
      "    -version           Display version information and exit\n",
      "\n",
      "Without refinement (very fast, avg accuracy similar to T-Coffee): -maxiters 2\n",
      "Fastest possible (amino acids): -maxiters 1 -diags -sv -distance1 kbit20_3\n",
      "Fastest possible (nucleotides): -maxiters 1 -diags\n"
     ]
    }
   ],
   "source": [
    "!muscle "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1056fa52-ab7f-4ce2-8e17-481baf292c15",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Run the aligner using mafft\n",
    "## It will save to the file SDHA_ncbi.mus.fasta \n",
    "\n",
    "!muscle -quiet -in SDHA_ncbi.fasta -out SDHA_ncbi.mus.fasta \n",
    "\n",
    "## Run the 'masking' or 'trimming' tool\n",
    "\n",
    "!trimal -gappyout -in SDHA_ncbi.mus.fasta  -out SDHA_ncbi.mus.go.fasta \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24866bba-69e7-45b8-beeb-d6d6c9f25784",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Run the tree making tool \n",
    "\n",
    "!iqtree -s SDHA_ncbi.mus.go.fasta --mset LG,WAG --mrate G,I -bb 1000 -pre SDHA_ncbi.mus.go.mfp -quiet\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
