{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "83b613fc-30c1-4bfb-af79-5dc46812ef1d",
   "metadata": {},
   "source": [
    "# **Day 2 - Part 2** \n",
    "\n",
    "## Learning objectives\n",
    "\n",
    "- Solve problems using python core\n",
    "- Introduction to biopython\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8df83589-4d04-44c8-aa2c-3a1324fb43b3",
   "metadata": {},
   "source": [
    "### Section 01 - Analyzing biological data with python core\n",
    "---\n",
    "\n",
    "We are going to start playing with actual biological data. \n",
    "\n",
    "[Here](https://perso.limsi.fr/pointal/_media/python:cours:mementopython3-english.pdf) you can find a useful cheat sheet with common python core functions.\n",
    "\n",
    "\n",
    "Let's reverse complement a DNA sequence. Hint, you can reverse strings using \"[::-1]\" or .reverse()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1807b073-17bc-4295-99c5-6e5eb8fcf247",
   "metadata": {},
   "outputs": [],
   "source": [
    "myDNA = \"agttgcccag\"\n",
    "RC_dict = {\"a\":\"t\",\"t\":\"a\",\"c\":\"g\",\"g\":\"c\"}\n",
    "myDNA_RC =\"\"\n",
    "for base in myDNA:\n",
    "    myDNA_RC += RC_dict[base]\n",
    "print(myDNA_RC)\n",
    "print(myDNA_RC[::-1])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "230c6997-ec06-4068-be80-bb7b3c00b0c0",
   "metadata": {},
   "source": [
    "Work on the following problems:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e12b5b49-0958-438f-b16e-6341af4a052e",
   "metadata": {},
   "source": [
    "### Rosalind Problem - Complementing a Strand of DNA\n",
    "\n",
    "Go on [Rosalind](http://rosalind.info/problems/revc/) website.\n",
    "\n",
    "> In DNA strings, symbols 'A' and 'T' are complements of each other, as are 'C' and 'G'.\n",
    "\n",
    "> The reverse complement of a DNA string 's' is the string 'sc' formed by reversing the symbols of 's', then taking the complement of each symbol (e.g., the reverse complement of \"GTCA\" is \"TGAC\").\n",
    "\n",
    "> Given: A DNA string 's' of length at most 1000 bp.\n",
    "\n",
    "> Return: The reverse complement 'sc' of 's'.\n",
    "\n",
    "Try to find other ways to reverse complement a sequence (without a dictionnary)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3ef1cfd-c010-4966-a4e4-65c68291620f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Correction\n",
    "\n",
    "#method1 using a for loop\n",
    "S = open('rosalind_revc_1_dataset.txt', 'r')\n",
    "S = S.read()\n",
    "\n",
    "revS = S[::-1]\n",
    "revS\n",
    "comp=' '\n",
    "for i in revS:\n",
    "    if i == 'A':              \n",
    "        comp+='T'             \n",
    "    elif i == 'C':            \n",
    "        comp+='G'             \n",
    "    elif i == 'T':            \n",
    "        comp+='A'             \n",
    "    elif i == 'G':            \n",
    "        comp+='C' \n",
    "print(revS)\n",
    "\n",
    "#method2 using a built-in function\n",
    "S = open('rosalind_revc_1_dataset.txt', 'r')\n",
    "S = S.read()\n",
    "\n",
    "revS = S[::-1]\n",
    "revS = revS.replace('A', 't')\n",
    "revS = revS.replace('T', 'a')\n",
    "revS = revS.replace('C', 'g')\n",
    "revS = revS.replace('G', 'c')\n",
    "revS = revS.upper()\n",
    "print(revS)\n",
    "\n",
    "#method2 simplified\n",
    "S = open('rosalind_revc_1_dataset.txt', 'r')\n",
    "S = S.read()\n",
    "\n",
    "S = S.replace('A', 't').replace('T', 'a').replace('C', 'g').replace('G', 'c').upper()[::-1]\n",
    "print(S)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d94a14e-9381-4d48-966f-2c933347abed",
   "metadata": {},
   "source": [
    "### Rosalind Problem - Counting DNA nucleotides\n",
    "\n",
    "Go on [Rosalind](http://rosalind.info/problems/dna/) website.\n",
    "\n",
    "> A string is simply an ordered collection of symbols selected from some alphabet and formed into a word\n",
    "\n",
    "> The length of a string is the number of symbols that it contains.\n",
    "\n",
    "> An example of a length 21 DNA string (whose alphabet contains the symbols 'A', 'C', 'G', and 'T') is \"ATGCTTCAGAAAGGTCTTACG.\"\n",
    "\n",
    "> Given: A DNA string 's' of length at most 1000 nt.\n",
    "\n",
    "> Return: Four integers (separated by spaces) counting the respective number of times that the symbols 'A', 'C', 'G', and 'T' occur in 's'.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebfe177f-65ea-444c-b3f5-ce4b6418d03d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Correction\n",
    "\n",
    "seq = open('rosalind_dna_1_dataset.txt', 'r')\n",
    "seq = seq.read()\n",
    "\n",
    "countA = 0\n",
    "countC = 0\n",
    "countG = 0\n",
    "countT = 0\n",
    "\n",
    "for n in seq:\n",
    "    if n == 'A':\n",
    "        countA += 1\n",
    "    if n == 'C':\n",
    "        countC += 1\n",
    "    if n == 'G':\n",
    "        countG += 1\n",
    "    if n == 'T':\n",
    "        countT += 1\n",
    "\n",
    "print(str(countA) + ' ' + str(countC) + ' ' + str(countG) + ' ' + str(countT))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9bf7fe4-2419-488a-8cd6-dcde39af298c",
   "metadata": {},
   "source": [
    "### Rosalind Problem - Transcribing DNA into RNA\n",
    "\n",
    "Go on [Rosalind](http://rosalind.info/problems/rna/) website.\n",
    "\n",
    "> An RNA string is a string formed from the alphabet containing 'A', 'C', 'G', and 'U'.\n",
    "\n",
    "> Given a DNA string 't' corresponding to a coding strand, its transcribed RNA string u is formed by replacing all occurrences of 'T' in t with 'U' in u\n",
    "\n",
    "> Given: A DNA string 't' having length at most 1000 nt.\n",
    "\n",
    "> Return: The transcribed RNA string of 't'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64473243-0487-4d18-a1c1-514b035a7dcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Correction\n",
    "\n",
    "#method1\n",
    "dna = open('rosalind_rna_1_dataset.txt', 'r')\n",
    "dna = dna.read()\n",
    "\n",
    "rna = dna.replace(\"T\", \"U\")\n",
    "print(rna)\n",
    "\n",
    "#method2\n",
    "dna = input()\n",
    "print(dna.replace(\"T\", \"U\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0647dfa8-0f9d-4210-a46a-31657f545fde",
   "metadata": {},
   "source": [
    "### Rosalind Problem - Counting Point Mutations\n",
    "\n",
    "Go on [Rosalind](http://rosalind.info/problems/hamm/) website.\n",
    "\n",
    "> Given two strings 's' and 't' of equal length, the Hamming distance between 's' and 't', denoted dH(s,t), is the number of corresponding symbols that differ in 's' and 't'\n",
    "\n",
    "> Given: Two DNA strings 's' and 't' of equal length (not exceeding 1 kbp).\n",
    "\n",
    "> Return: The Hamming distance dH(s,t).\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "705196f7-a5b0-4cde-9627-835d45746708",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Correction\n",
    "\n",
    "#method1 with a while loop\n",
    "a = input()\n",
    "b= input()\n",
    "n = 0\n",
    "hamming = 0\n",
    "while n <= len(a)-1:\n",
    "    if a[n] != b[n]:\n",
    "        hamming += 1\n",
    "    n += 1\n",
    "    \n",
    "print(hamming)\n",
    "\n",
    "#method2 with a function\n",
    "def hamming(a, b):\n",
    "    return len(list(filter(lambda pair: pair[0]!=pair[1], zip(a,b))))\n",
    "\n",
    "hamming(a,b) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8078d01-c3b9-4d7b-abff-56bd8a6348f2",
   "metadata": {},
   "source": [
    "### Rosalind Problem - Finding a Motif in DNA\n",
    "\n",
    "Go on [Rosalind](http://rosalind.info/problems/subs/) website.\n",
    "\n",
    "> Given two strings 's' and 't', 't' is a substring of 's' if 't' is contained as a contiguous collection of symbols in 's' (as a result, t must be no longer than s).\n",
    "\n",
    "> The position of a symbol in a string is the total number of symbols found to its left, including itself (e.g., the positions of all occurrences of 'U' in \"AUGCUUCAGAAAGGUCUUACG\" are 2, 5, 6, 15, 17, and 18). \n",
    "\n",
    "> The symbol at position i of 's' is denoted by s[i]\n",
    "\n",
    "> A substring of s can be represented as s[j:k], where j and k represent the starting and ending positions of the substring in s; \n",
    "\n",
    "> for example, if s = \"AUGCUUCAGAAAGGUCUUACG\", then s[2:5] = \"UGCU\".\n",
    "\n",
    "> The location of a substring s[j:k] is its beginning position j; note that t will have multiple locations in 's' if it occurs more than once as a substring of 's' \n",
    "\n",
    "> Given: Two DNA strings 's' and 't' (each of length at most 1 kbp).\n",
    "\n",
    "> Return: All locations of 't' as a substring of 's'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e11e888-5757-4e0c-9152-83572f4e9546",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Correction\n",
    "s1,s2 = open('rosalind_subs_1_dataset.txt').read().split()\n",
    "print(s1)\n",
    "print(s2)\n",
    "\n",
    "loc = []\n",
    "for i in range(len(s1)):\n",
    "    if s1[i:].startswith(s2):\n",
    "        print(i+1)\n",
    "        loc.append(i+1)\n",
    "\n",
    "print(loc)\n",
    "        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a09f48ec-1cc5-4ef5-83ce-feee6149c1dc",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Section 02 - Biopython\n",
    "---\n",
    "\n",
    "When starting to code, chances are you on not the first person to try and write code to solve that problem. Some very clever biologists wrote a package called biopython that has most of the functionality one would ever need in a python program. \n",
    "\n",
    "[See the cookbook for biopython](http://biopython.org/DIST/docs/tutorial/Tutorial.html)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07deb7f8-c5ca-4253-8c7a-444e9a62fe73",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check the installation and path\n",
    "from Bio import SeqIO\n",
    "print(SeqIO.__file__)\n",
    "\n",
    "# if biopython not installed\n",
    "# pip install biopython\n",
    "# if no pip\n",
    "# python -m ensurepip --upgrade"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9954241f-9431-481a-902a-7a7961eef311",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For example\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "# we can convert a string into a sequence object using Seq()\n",
    "myDNA = Seq(\"agttgcccag\")\n",
    "\n",
    "# Use the function 'reverse_complement()'\n",
    "print(myDNA.reverse_complement())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e91605ca-5c69-4fd6-8922-5645a8a3b830",
   "metadata": {},
   "source": [
    "### Rosalind Problem - Transcribing DNA into RNA\n",
    "\n",
    "Repeat this [Rosalind](http://rosalind.info/problems/rna/) exercice using a biopython function.\n",
    "\n",
    "> An RNA string is a string formed from the alphabet containing 'A', 'C', 'G', and 'U'.\n",
    "\n",
    "> Given a DNA string 't' corresponding to a coding strand, its transcribed RNA string u is formed by replacing all occurrences of 'T' in t with 'U' in u\n",
    "\n",
    "> Given: A DNA string 't' having length at most 1000 nt.\n",
    "\n",
    "> Return: The transcribed RNA string of 't'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40dbdfa0-bbc8-40f6-94fc-1c60b2b47991",
   "metadata": {},
   "outputs": [],
   "source": [
    "dna = open('rosalind_rna_1_dataset.txt', 'r').read()\n",
    "#dna = dna.read()\n",
    "dna = Seq(dna)\n",
    "\n",
    "#dna = Seq(\"GTGCTTAGACACTACCCTGCTGCTACGTTTGTTACGACGTGTCTATACCTGTTATCTAGACTTTACAAACCGAAGTAGGCCAACATTTGCGAAATTTAAAGTGACGAGTGCCAG\")\n",
    "print(dna)\n",
    "print(dna.transcribe())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d814aa76-1626-4d42-bee4-7299b36dc074",
   "metadata": {},
   "source": [
    "To convert between DNA and RNA we can use the `transcribe()` function and to convert between DNA/RNA and amino acids we can use the `translate()` function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aae8aad0-9318-4a59-b2a7-ec7b3a4edca7",
   "metadata": {},
   "outputs": [],
   "source": [
    "myDNA = Seq(\"agtcaatcccaacgggagccccag\")\n",
    "\n",
    "\n",
    "myRNA = myDNA.transcribe()\n",
    "myProtein = myDNA.translate()\n",
    "\n",
    "print(myRNA)\n",
    "print(myProtein)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a916c2b-90f8-4d3a-9a2c-e9b73e16445d",
   "metadata": {},
   "source": [
    "### Reading fasta files - Python Core\n",
    "\n",
    "One of the most common methods to store DNA and protein information is in FASTA files. The consist of two main elements: \n",
    "\n",
    "- the title of the sequence (also called header) preceded by a '>'\n",
    "- the sequence\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "d1f492c1-fc32-49ed-a271-236b4a428ec3",
   "metadata": {},
   "source": [
    ">Wrapped_seq\n",
    "GGGAAAGGAAGGGGAAAAAAAAATTTTAAAGGAATTAAAGGAATGGAATTTAAAGGAGGAAATTTAAAAGAAAGAAGGGAATTGGTAAGGAAAGGAAAGGAA\n",
    "\n",
    ">MultiLine_seq\n",
    "GGGAATTTGGGGAAATTTAAAAGAAAGAAGGGAATTTGGGGAAATTTAAAAGAAAGAA\n",
    "GGAATTTAAAGGAGGAAATTTAAAAGAAAGAAGGGAATTGGTAAGGAAAGGAAAGGAA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c650d63-b087-4db5-80c9-62ffe2c57324",
   "metadata": {},
   "source": [
    "Notice how the the sequence can be wrapped or multiple lines? This can be frustrating whey trying to read fasta files using the core python functions.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "884b713c-6308-4c5b-8bed-9e563ba2eb7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's try and make a dictionary storing \n",
    "# the sequence title as a key and the sequence as the value\n",
    "seqDict={}\n",
    "myLines = open(\"MyFastaLines.txt\").readlines()\n",
    "\n",
    "for line in myLines:\n",
    "    # you can see there are /n characters? these are 'end-of-line characters'\n",
    "    print(line)  \n",
    "    # we can remove these using '.strip()' function\n",
    "    l = line.strip()\n",
    "    if  l.startswith(\">\"):\n",
    "        #Easy, if the line starts with a > then it must be the header\n",
    "        titleline= l\n",
    "        seqDict[titleline]=\"\"\n",
    "    else:\n",
    "        # The next line must be part of the sequence\n",
    "        seqDict[titleline]= l\n",
    "        \n",
    "# does this look right?\n",
    "print(seqDict)\n",
    "\n",
    "# How you could fix this code? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8f5a084-3750-4e62-9898-d62a584c3f58",
   "metadata": {},
   "outputs": [],
   "source": [
    "# seqDict[titleline]+= l\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ead6fa1-edb3-41de-b8f4-a0fccd4006f7",
   "metadata": {},
   "source": [
    "### Writing fasta files - Python Core\n",
    "\n",
    "You can create a fasta file using python core from variables of your choosing. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0510dd31-05ec-45a8-8fc3-73732837dafd",
   "metadata": {},
   "outputs": [],
   "source": [
    "seqDict = {\"seq1\":\"AGTCGTA\", \"seq2\":\"GTCACTGC\"}\n",
    "\n",
    "#one way\n",
    "with open(\"myoutputSeq.fasta\",\"w\") as output:\n",
    "    for seq in seqDict:\n",
    "        output.write(\">\" + seq + \"\\n\" + seqDict[seq] + \"\\n\")\n",
    "        \n",
    "#another way using % operator\n",
    "with open(\"myoutputSeq.fasta\",\"w\") as output:\n",
    "    for seq in seqDict:\n",
    "        # %s operator is read as a placeholder for the strings in the parentheses \n",
    "        # after the string\n",
    "        output.write(\">%s\\n%s\\n\" % (seq, seqDict[seq]))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0001565-8188-46cf-9b23-b4f73914e004",
   "metadata": {},
   "source": [
    "### Biopython and Fasta files \n",
    "\n",
    "Of course, there is a much easier way to read and write fasta files using biopython using the `SeqIO` module. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd10ed6b-0f38-4be5-8957-bb89da4e104d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "# SeqIO.read takes two arguments, the file and the format of the file\n",
    "# you can use 'read' when you file has one sequence\n",
    "record = SeqIO.read(\"myFavouriteProtein_ncbi.fasta\", 'fasta')\n",
    "\n",
    "#Here is some information available in the 'record' object\n",
    "header = record.description\n",
    "id = record.id\n",
    "sequence = record.seq\n",
    "\n",
    "print(header)\n",
    "print(id)\n",
    "print(sequence)\n",
    "\n",
    "# You can write this record to another file\n",
    "SeqIO.write(record, \"myFavouriteProtein_ncbi_copy.fasta\", \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d33d1de-6f64-4f56-9e2f-1941e92b0958",
   "metadata": {},
   "outputs": [],
   "source": [
    "# You can read a file with more than one sequence using  SeqIO.parse()\n",
    "outputList = []\n",
    "for record in SeqIO.parse(\"NFU1_proteins.fasta\", 'fasta'):\n",
    "    outputList.append(record)\n",
    "    # however, to write more than one sequence to a file using SeqIO, \n",
    "    # you have to save the record objects in a list and write them later\n",
    "\n",
    "SeqIO.write(outputList, \"NFU1_proteins_copy.fasta\", \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1516dacf-9bfd-455a-894f-f6cfc7c8f265",
   "metadata": {},
   "outputs": [],
   "source": [
    "# HAND IN #5\n",
    "\n",
    "# Write a fasta parser for the fastafile 'NFU1_proteins.fasta'\n",
    "# Use this parser to indicate how many sequences are in the file\n",
    "\n",
    "# make a file with the longest sequence called 'NFU1_longestSeq.fasta'\n",
    "\n",
    "# save the lengths of all the proteins as a list 'all_seq_lengths'\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32bd4009-ecfe-4d13-867f-90ef53c3bf6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "maxSeqs=[]\n",
    "maxSeq_length = 0\n",
    "all_seq_lengths = []\n",
    "\n",
    "for record in SeqIO.parse(\"NFU1_proteins.fasta\", 'fasta'):\n",
    "    all_seq_lengths.append(len(record.seq))\n",
    "    if len(record.seq) > maxSeq_length:\n",
    "        # wipe the list! we found a new winner!\n",
    "        maxSeqs=[]\n",
    "        maxSeq_length = len(record.seq)\n",
    "        maxSeqs.append(record)\n",
    "    elif len(record.seq) == maxSeq_length:\n",
    "        # we found something the same lenght as the best, add it to the list!\n",
    "        maxSeqs.append(record)\n",
    "        \n",
    "print(\"There are %i sequences in the file\" % (len(all_seq_lengths)))\n",
    "\n",
    "print(all_seq_lengths)\n",
    "\n",
    "SeqIO.write(maxSeqs, \"NFU1_longestSeq.fasta\", \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1bade67-8a41-4363-9e91-6e35b6c2693b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
